src/notifications/NotificationsManager.ts

```diff
@@ -161,26 +353,210 @@ class NotificationsManager {
+    const transactionCallback = async (tran: DBTransaction) => {
+      await tran.put(
+        [
+          ...this.notificationsManagerOutboxNotificationsDbPath,
+          idUtils.toBuffer(notificationId),
+        ],
+        notificationDb,
+      );
+      const counter = await tran.get<number>(
+        this.notificationsManagerOutboxNotificationsCounterDbPath,
+      );
+      await tran.put(
+        this.notificationsManagerOutboxNotificationsCounterDbPath,
+        counter != null ? counter + 1 : 1,
+      );
+      pendingTask = await this.taskManager.scheduleTask(
+        {
+          handlerId: this.sendNotificationHandlerId,
+          parameters: [
+            {
+              nodeIdEncoded,
+              notificationIdEncoded,
+              retries,
+              attempt: 0,
+            },
+          ],
+          path: [this.sendNotificationHandlerId],
+          lazy: false,
+        },
+        tran,
+      );
+      this.outboxNotificationIdEncodedToTaskIdMap.set(
+        notificationIdEncoded,
+        pendingTask.id,
+      );
+    };
+    if (tran == null) {
+      await this.db.withTransactionF(transactionCallback);
+    } else {
+      await transactionCallback(tran);
+    }
+    const sendProm = (async () => {
+      while (pendingTask != null) {
+        pendingTask = await pendingTask.promise();
+      }
+    })();
+    sendProm.catch(() => {});
+    return {
+      notificationId,
+      sendProm,
+    };
+  }
+
+  protected async getOutboxNotificationIds(
+    tran: DBTransaction,
+  ): Promise<Array<NotificationId>> {
+    const notificationIds: Array<NotificationId> = [];
+    const messageIterator = tran.iterator<NotificationDB>(
+      this.notificationsManagerOutboxNotificationsDbPath,
+      { valueAsBuffer: false },
+    );
+    for await (const [keyPath] of messageIterator) {
+      const key = keyPath[0] as Buffer;
+      const notificationId = IdInternal.fromBuffer<NotificationId>(key);
+      notificationIds.push(notificationId);
+    }
+    return notificationIds;
+  }
+
+  protected async readOutboxNotificationById(
+    notificationId: NotificationId,
+    tran: DBTransaction,
+  ): Promise<Notification | undefined> {
+    const notificationDb = await tran.get<NotificationDB>([
+      ...this.notificationsManagerOutboxNotificationsDbPath,
+      idUtils.toBuffer(notificationId),
+    ]);
+    if (notificationDb == null) {
+      return undefined;
+    }
+    return {
+      notificationIdEncoded:
+        notificationsUtils.encodeNotificationId(notificationId),
+      ...notificationDb,
+    };
+  }
+
+  /**
+   * Read pending notifications in the outbox.
+   */
+  @ready(new notificationsErrors.ErrorNotificationsNotRunning())
+  public async readOutboxNotifications({
+    number = 'all',
+    order = 'newest',
+    tran,
+  }: {
+    unread?: boolean;
+    number?: number | 'all';
+    order?: 'newest' | 'oldest';
+    tran?: DBTransaction;
+  } = {}): Promise<Array<Notification>> {
```

Comment:
Likewise, this should be a generator.

Reply:
should i change the readNotifications fn to also be an asyncgenerator?

Reply:
Yes.